File: ios/Sources/KlinikAPI/FollowUpAPI.swift
import Foundation
import KlinikCore

public enum MilestoneStatus: String, Decodable, Encodable, Sendable, Equatable {
    case pending = "PENDING"
    case notified = "NOTIFIED"
    case completed = "COMPLETED"
    case missed = "MISSED"
    case skipped = "SKIPPED"

    public var localizedName: String { L10n.string("followUp.status.\(rawValue)") }

    /// Whether the clinic is still waiting for this visit to happen.
    public var isOutstanding: Bool { self == .pending || self == .notified || self == .missed }
}

public struct Milestone: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    /// D1, W1, M1… — the same labels the photo gallery uses for its phases.
    public let label: String
    public let dueAt: Date
    public let status: MilestoneStatus
    public let notifiedAt: Date?
    public let completedAt: Date?

    public var localizedLabel: String { L10n.string("followUp.milestone.\(label)") }
}

public struct FollowUpSchedule: Decodable, Sendable, Equatable {
    public let id: String
    public let patientId: String
    public let surgeryDate: Date
    public let template: String?
    /// Soonest first.
    public let milestones: [Milestone]

    public init(
        id: String,
        patientId: String,
        surgeryDate: Date,
        template: String?,
        milestones: [Milestone]
    ) {
        self.id = id
        self.patientId = patientId
        self.surgeryDate = surgeryDate
        self.template = template
        self.milestones = milestones
    }

    /// The next visit still ahead, which is what a patient opens this for.
    public func next(after now: Date = Date()) -> Milestone? {
        milestones.first { $0.dueAt >= now && $0.status.isOutstanding }
    }

    public var missed: [Milestone] { milestones.filter { $0.status == .missed } }
}

/// The endpoint answers with an empty object when no schedule exists yet.
private struct OptionalSchedule: Decodable, Sendable {
    let value: FollowUpSchedule?

    init(from decoder: Decoder) throws {
        value = try? FollowUpSchedule(from: decoder)
    }
}

public struct FollowUpAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func mine() async throws -> FollowUpSchedule? {
        try await client.send(Endpoint(method: .get, path: "me/follow-up"), as: OptionalSchedule.self)
            .value
    }

    public func forPatient(_ patientId: String) async throws -> FollowUpSchedule? {
        try await client.send(
            Endpoint(method: .get, path: "patients/\(patientId)/follow-up"),
            as: OptionalSchedule.self
        ).value
    }

    public func generate(
        patientId: String,
        surgeryDate: Date,
        template: String? = nil,
        timezone: String? = nil
    ) async throws -> FollowUpSchedule {
        try await client.send(
            Endpoint(
                method: .post,
                path: "patients/\(patientId)/follow-up",
                body: try JSONEncoder.klinik.encode(
                    GenerateBody(surgeryDate: surgeryDate, template: template, timezone: timezone)
                )
            ),
            as: FollowUpSchedule.self
        )
    }

    public func setStatus(milestoneId: String, status: MilestoneStatus) async throws -> Milestone {
        try await client.send(
            Endpoint(
                method: .patch,
                path: "follow-up/milestones/\(milestoneId)",
                body: try JSONEncoder.klinik.encode(StatusBody(status: status))
            ),
            as: Milestone.self
        )
    }

    private struct GenerateBody: Encodable {
        let surgeryDate: Date
        let template: String?
        let timezone: String?
    }

    private struct StatusBody: Encodable {
        let status: MilestoneStatus
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/FollowUpApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class MilestoneStatus {
    PENDING,
    NOTIFIED,
    COMPLETED,
    MISSED,
    SKIPPED,
    ;

    val stringKey: String get() = "followup_status_${name.lowercase()}"

    /** Whether the clinic is still waiting for this visit to happen. */
    val isOutstanding: Boolean get() = this == PENDING || this == NOTIFIED || this == MISSED
}

@Serializable
data class Milestone(
    val id: String,
    /** D1, W1, M1… — the same labels the photo gallery uses for its phases. */
    val label: String,
    val dueAt: String,
    val status: MilestoneStatus = MilestoneStatus.PENDING,
    val notifiedAt: String? = null,
    val completedAt: String? = null,
) {
    val stringKey: String get() = "followup_milestone_${label.lowercase()}"
}

@Serializable
data class FollowUpSchedule(
    val id: String,
    val patientId: String,
    val surgeryDate: String,
    val template: String? = null,
    /** Soonest first. */
    val milestones: List<Milestone> = emptyList(),
) {
    /**
     * The next visit still ahead, which is what a patient opens this for.
     *
     * Compared as ISO-8601 strings, which sort chronologically when they are
     * all UTC — and the server always sends UTC. Parsing would need a date
     * library in a module that has none, for an ordering the format already
     * gives.
     */
    fun next(nowIso: String): Milestone? =
        milestones.firstOrNull { it.dueAt >= nowIso && it.status.isOutstanding }

    val missed: List<Milestone> get() = milestones.filter { it.status == MilestoneStatus.MISSED }
}

@Serializable
private data class GenerateBody(
    val surgeryDate: String,
    val template: String? = null,
    val timezone: String? = null,
)

@Serializable
private data class StatusBody(val status: MilestoneStatus)

class FollowUpApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /** Null when no schedule has been generated: the endpoint answers `{}`. */
    suspend fun mine(): FollowUpSchedule? =
        optional(client.send(Endpoint(HttpMethod.GET, "me/follow-up")))

    suspend fun forPatient(patientId: String): FollowUpSchedule? =
        optional(client.send(Endpoint(HttpMethod.GET, "patients/$patientId/follow-up")))

    suspend fun generate(
        patientId: String,
        surgeryDate: String,
        template: String? = null,
        timezone: String? = null,
    ): FollowUpSchedule =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.POST,
                    "patients/$patientId/follow-up",
                    body = json.encodeToString(
                        GenerateBody.serializer(),
                        GenerateBody(surgeryDate, template, timezone),
                    ),
                ),
            ),
        )

    suspend fun setStatus(milestoneId: String, status: MilestoneStatus): Milestone =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.PATCH,
                    "follow-up/milestones/$milestoneId",
                    body = json.encodeToString(StatusBody.serializer(), StatusBody(status)),
                ),
            ),
        )

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }

    private fun optional(body: String): FollowUpSchedule? =
        runCatching { json.decodeFromString<FollowUpSchedule>(body) }.getOrNull()
}
